src/components/utils/src/timer.cc

```diff
@@ -173,46 +135,137 @@ void timer::Timer::OnTimeout() {
   LOG4CXX_DEBUG(logger_,
                 "Timer has finished counting. Timeout(ms): "
                     << static_cast<uint32_t>(timeout_ms_));
-  {
-    // Task locked by own lock because from this task in callback we can
-    // call Stop of this timer and get DeadLock
-    sync_primitives::AutoLock auto_lock(task_lock_);
-    DCHECK(task_);
-    task_->run();
-  }
+
+  // Task locked by own lock because from this task in callback we can
+  // call Stop of this timer and get DeadLock
+  sync_primitives::AutoLock auto_lock(task_lock_);
+  DCHECK(task_);
+  task_->run();
+}
+
+timer::Milliseconds timer::Timer::GetTimeout() const {
   sync_primitives::AutoLock auto_lock(lock_);
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
+  return timeout_ms_;
+}
+
+timer::Timer::TimerDelegate::TimerDelegate(Timer* timer)
+    : timer_(timer)
+    , timeout_milliseconds_(0)
+    , state_lock_(true)
+    , stop_flag_(false)
+    , restart_flag_(false)
+    , is_started_flag_(false) {
+  DCHECK(timer_);
+}
+
+timer::Timer::TimerLooperDelegate::TimerLooperDelegate(timer::Timer* timer)
+    : TimerDelegate(timer) {}
+
+timer::Timer::TimerDelegate::~TimerDelegate() {
+  timer_ = NULL;
+}
+
+void timer::Timer::TimerDelegate::threadMain() {
+  using sync_primitives::ConditionalVariable;
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  stop_flag_ = false;
+  // Notify that we've already started
+  TimerDelegate::is_started_flag_ = true;
+  TimerDelegate::starting_condition_.NotifyOne();
+
+  while (!stop_flag_) {
+    // Sleep
+    int32_t wait_milliseconds_left = TimerDelegate::Get_timeout();
+    LOG4CXX_DEBUG(logger_,
+                  "Milliseconds left to wait: " << wait_milliseconds_left);
+    ConditionalVariable::WaitStatus wait_status =
+        termination_condition_.WaitFor(auto_lock, wait_milliseconds_left);
+    // Quit sleeping or continue sleeping in case of spurious wake up
+    if (ConditionalVariable::kTimeout == wait_status ||
+        wait_milliseconds_left <= 0) {
+      LOG4CXX_DEBUG(logger_,
+                    "Timer has finished counting. Timeout(ms): "
+                        << wait_milliseconds_left);
+      timer_->OnTimeout();
+    } else {
+      LOG4CXX_DEBUG(
+          logger_,
+          "Timeout reset force (ms): " << TimerDelegate::timeout_milliseconds_);
+    }
+    {
+      sync_primitives::AutoLock auto_lock(restart_flag_lock_);
+      if (!restart_flag_) {
+        return;
+      }
+      restart_flag_ = false;
+    }
   }
-  if (repeatable_) {
-    StartUnsafe();
+}
+
+void timer::Timer::TimerLooperDelegate::threadMain() {
+  using sync_primitives::ConditionalVariable;
+  sync_primitives::AutoLock auto_lock(TimerDelegate::state_lock_);
+  TimerDelegate::stop_flag_ = false;
+  // Notify that we've already started
+  is_started_flag_ = true;
+  starting_condition_.NotifyOne();
+
+  while (!TimerDelegate::stop_flag_) {
```

Comment:
There is at leaset two possible ways how to avoid this duplication. So please do that